### PR TITLE
Comply with error handling policy in Tracing API

### PR DESCRIPTION
**Pull request by lmolkova**
First chunk of https://github.com/open-telemetry/opentelemetry-dotnet/issues/295

This change brings a few things:
- Span implementations never throw in runtime
- Tracer implementation never throw in runtime
- propagators never throw in runtime
- resource creation: does not throw

There is more coming for metrics and distributed context

Fallback for invalid args (**let's discuss!**)
1. if span Name (event description, attribute name or value) is null, we fallback to empty string (just to make exporters life a bit easier), updating span name to null - would update to empty string
2. Setting invalid status is noop (keeps default) 
3. If event in AddEvent is null, we don't add it
4. If link is created with null or invalid context - we still create it and add to span

The main alternative I see is to **not add links/events/attributes that are not valid**. What stops me from doing it: finding root cause when attribute/event name/span name/etc is invalid is obvious if we still populate them and much more complicated if we swallow it.

Please share your thoughts.

### CODE REVIEW

File: src/OpenTelemetry.Api/Context/Propagation/TraceContextFormat.cs [hunk 1]
```diff
@@ -81,6 +82,23 @@ public SpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> ge
         /// <inheritdoc/>
         public void Inject<T>(SpanContext spanContext, T carrier, Action<T, string, string> setter)
         {
+            if (spanContext == null || !spanContext.IsValid)
+            {
+                OpenTelemetryApiEventSource.Log.FailedToInjectSpanContext("Invalid context");
+            }
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Api/Context/Propagation/TraceContextFormat.cs]: Should we return early when the spanContext is invalid like we do if carrier and setter are null?

File: src/OpenTelemetry.Api/Context/Propagation/BinaryFormat.cs [hunk 2]
```diff
@@ -43,17 +44,19 @@ public class BinaryFormat : IBinaryFormat
...
 
+        private static readonly byte[] InvalidContext = new byte[0];
+
         /// <inheritdoc />
         public SpanContext FromByteArray(byte[] bytes)
         {
             if (bytes == null)
             {
-                throw new ArgumentNullException(nameof(bytes));
+                return SpanContext.BlankRemote;
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Api/Context/Propagation/BinaryFormat.cs]: We could emit logs if bytes is not the expected length or if bytes[0] is not VersionId.
eg `OpenTelemetryApiEventSource.Log.InvalidSpanContextBytes("null bytes");`

File: src/OpenTelemetry.Api/Internal/BlankSpan.cs [hunk 3]
```diff
@@ -39,108 +39,51 @@ private BlankSpan()
...
-                {
-                    throw new ArgumentException(nameof(value));
-                }
-            }
-        }
+        public Status Status { get; set; } = Status.Ok;
 
         /// <inheritdoc />
         public void UpdateName(string name)
         {
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Api/Internal/BlankSpan.cs]: For consistency, we could also log null values for these functions.

File: src/OpenTelemetry.Api/Trace/ProxyTracer.cs [hunk 4]
```diff
@@ -50,51 +50,21 @@ public IDisposable WithSpan(ISpan span, bool endOnDispose)
 
         public ISpan StartRootSpan(string operationName, SpanKind kind, SpanCreationOptions options)
         {
-            if (operationName == null)
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Api/Trace/ProxyTracer.cs]: Could these functions also emit logs when values are null?

File: src/OpenTelemetry/Context/Propagation/B3Format.cs [hunk 5]
```diff
@@ -76,7 +77,7 @@ public SpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> ge
                 }
                 else
                 {
-                    throw new SpanContextParseException("Missing X_B3_TRACE_ID.");
+                    return SpanContext.BlankRemote;
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Context/Propagation/B3Format.cs]: Same, could we emit log when failing to parse span context is invalid?

File: src/OpenTelemetry/Trace/Span.cs [hunk 6]
```diff
@@ -204,22 +213,12 @@ public void UpdateName(string name)
...
             }
 
-            this.Name = name ?? throw new ArgumentNullException(nameof(name));
+            this.Name = name ?? string.Empty;
         }
 
         /// <inheritdoc/>
         public void SetAttribute(KeyValuePair<string, object> keyValuePair)
         {
-            if (keyValuePair.Key == null)
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Trace/Span.cs]: Emit logs when failing to add attribute.

File: src/OpenTelemetry/Trace/Tracer.cs [hunk 7]
```diff
@@ -69,9 +69,9 @@ internal Tracer(SpanProcessor spanProcessor, Sampler sampler, TracerConfiguratio
 
         public IDisposable WithSpan(ISpan span, bool endSpanOnDispose)
         {
-            if (span == null)
-            {
-                throw new ArgumentNullException(nameof(span));
+            if (span != null)
+            {
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Trace/Tracer.cs]: whitespace

File: src/OpenTelemetry/Trace/Tracer.cs [hunk 8]
```diff
@@ -85,22 +85,12 @@ public IDisposable WithSpan(ISpan span, bool endSpanOnDispose)
         /// <inheritdoc/>
         public ISpan StartRootSpan(string operationName, SpanKind kind, SpanCreationOptions options)
         {
-            if (operationName == null)
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Trace/Tracer.cs]: Emit logs when starting new root span with invalid params.

File: src/OpenTelemetry/Context/Propagation/B3Format.cs (hunk 5 above)
Comment (lmolkova) [reply src/OpenTelemetry/Context/Propagation/B3Format.cs]: This branch represents the case where there is no header in the message with b3 (this is a happy and default case). I don't think any logging needs to happen here.

File: src/OpenTelemetry/Trace/Span.cs (hunk 6 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Span.cs]: we are not failing to add it, we add it anyway, but with empty value

File: src/OpenTelemetry/Trace/Tracer.cs (hunk 7 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Tracer.cs]: ? what is the problem here?

File: src/OpenTelemetry/Trace/Tracer.cs (hunk 8 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Tracer.cs]: done in span constructor

File: src/OpenTelemetry.Api/Internal/BlankSpan.cs (hunk 3 above)
Comment (lmolkova) [reply src/OpenTelemetry.Api/Internal/BlankSpan.cs]: I've done it on the real span, validation makes little sense on blank one

File: src/OpenTelemetry.Api/Trace/ProxyTracer.cs (hunk 4 above)
Comment (lmolkova) [reply src/OpenTelemetry.Api/Trace/ProxyTracer.cs]: done it on the span impl, so we don't have to repeat it so many times

File: src/OpenTelemetry/Trace/Tracer.cs (hunk 7 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Tracer.cs]: oh, I see. I'm ok to fix it, but I'd prefer it either be enforced by the tooling or not enforced at all.

Comment (MikeGoldsmith) [reply src/OpenTelemetry/Trace/Tracer.cs]: #372 includes a setting in .editorconfig to remove trailing whitespace, which will automatically be applied once merged.

**Pull request merged**